File: apps/api/src/http/handler/service.ts
// The probe and the attach socket. Both public: the socket is gated by its own token,
// not the service token.
import { Hono } from 'hono'
import { Err } from '@sandboxd/core/errors'
import type { Tokens } from '../../tokens'
import type { AttachData } from '../../attach'
import { AttachQuery, Ok } from '../../schema'
import { Doc } from '../doc'

/** What Bun.serve hands `app.fetch` as the second argument. */
export interface Upgrader {
  upgrade(req: Request, opts: { data: AttachData }): boolean
}
export type Bindings = { Bindings: Upgrader }

export function serviceRoutes(tokens: Tokens) {
  const doc = Doc.tag('Service')
  return new Hono<Bindings>()
    .get('/healthz', doc('Liveness', { 200: Doc.json('Up.', Ok) }), (c) =>
      c.json({ ok: true } as const, 200),
    )
    .get(
      '/attach',
      doc(
        'Attach to a session terminal',
        {
          101: { description: 'WebSocket: binary frames are PTY bytes both ways.' },
          ...Doc.errors(401, 426),
        },
        'Browser endpoint. `token` comes from POST /sessions/:id/attach-token.',
      ),
      Doc.validator('query', AttachQuery),
      (c) => {
        const q = c.req.valid('query')
        const p = tokens.verify(q.token, 'attach')
        if (!p) throw Err.unauthorized('invalid or expired attach token')
        const data: AttachData = {
          kind: 'attach',
          sid: p.sid,
          pty: null,
          cols: q.cols,
          rows: q.rows,
        }
        return c.env.upgrade(c.req.raw, { data })
          ? new Response(null, { status: 101 })
          : c.json({ error: 'websocket upgrade required' }, 426)
      },
    )
}

File: internal/cp/http/service.go
package http

import (
	"context"
	"net/http"

	"github.com/danielgtaylor/huma/v2"
)

// The probe, and the document entry for the socket huma does not serve.

func registerService(api huma.API) {
	huma.Register(api, huma.Operation{
		OperationID: "healthz",
		Method:      http.MethodGet,
		Path:        "/healthz",
		Summary:     "Liveness",
		Tags:        []string{"Service"},
		Security:    []map[string][]string{},
	}, func(_ context.Context, _ *struct{}) (*okOut, error) {
		return &okOut{Body: ok{OK: true}}, nil
	})
}

// documentAttach writes the attach socket into the document by hand. It is a WebSocket
// upgrade with a hijacked connection, so it is not a huma operation — but a client
// reading the document still needs to find it.
func documentAttach(doc *huma.OpenAPI) {
	str := &huma.Schema{Type: "string"}
	num := &huma.Schema{Type: "integer"}
	doc.Paths["/attach"] = &huma.PathItem{
		Get: &huma.Operation{
			OperationID: "attach",
			Summary:     "Attach to a sandbox terminal",
			Description: "Browser endpoint. `token` comes from POST /sandboxes/{id}/attach-token. " +
				"Binary frames are raw PTY bytes both ways; text frames are JSON — the client " +
				"sends {type:\"resize\",cols,rows}, the server sends {type:\"closed\",reason}.",
			Tags:     []string{"Service"},
			Security: []map[string][]string{},
			Parameters: []*huma.Param{
				{Name: "token", In: "query", Required: true, Schema: str},
				{Name: "cols", In: "query", Schema: num, Description: "Default 120."},
				{Name: "rows", In: "query", Schema: num, Description: "Default 40."},
			},
			Responses: map[string]*huma.Response{
				"101": {Description: "WebSocket: binary frames are PTY bytes both ways."},
				"401": {Description: "Invalid or expired attach token."},
			},
		},
	}
}
